File: 11_GLKit_三角形变换_OC/08_GLKit_OC/CustomViewController.m
//
//  CustomViewController.m
//  08_GLKit_OC
//
//  Created by 陈嘉琳 on 2020/8/1.
//  Copyright © 2020 CJL. All rights reserved.
//

#import "CustomViewController.h"
#import <OpenGLES/ES2/gl.h>

@interface CustomViewController ()

@property (nonatomic, strong) EAGLContext *mContext;
@property (nonatomic, strong) GLKBaseEffect *mEffect;

@property (nonatomic, assign) int count;

//旋转的度数
@property (nonatomic, assign) float xDegree;
@property (nonatomic, assign) float yDegree;
@property (nonatomic, assign) float zDegree;

//是否旋转x，y，z
@property (nonatomic, assign) BOOL bX;
@property (nonatomic, assign) BOOL bY;
@property (nonatomic, assign) BOOL bZ;

@end

@implementation CustomViewController

{
    dispatch_source_t timer;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
//    self.view.backgroundColor = [UIColor blackColor];
    
    //1.新建图层
    [self setupContext];
    
    //2.渲染图形
    [self render];
    
}



- (void) setupContext{

    //1.新建OpenGL ES上下文
    self.mContext = [[EAGLContext alloc] initWithAPI:kEAGLRenderingAPIOpenGLES2];
    
    GLKView *view = (GLKView*)self.view;
    view.context = self.mContext;
    view.drawableColorFormat = GLKViewDrawableColorFormatRGBA8888;
    view.drawableDepthFormat = GLKViewDrawableDepthFormat24;
    
    [EAGLContext setCurrentContext:self.mContext];
    
    glEnable(GL_DEPTH_TEST);
    
}

//渲染图形
- (void)render{
    
    //1.顶点数据
       //前3个元素，是顶点数据；中间3个元素，是顶点颜色值，最后2个是纹理坐标
    GLfloat attrArr[] =
    {
        -0.5f, 0.5f, 0.0f,      1.0f, 0.0f, 1.0f, //左上
        0.5f, 0.5f, 0.0f,       1.0f, 0.0f, 1.0f, //右上
        -0.5f, -0.5f, 0.0f,     1.0f, 1.0f, 1.0f, //左下
        
        0.5f, -0.5f, 0.0f,      1.0f, 1.0f, 1.0f, //右下
        0.0f, 0.0f, 1.0f,       0.0f, 1.0f, 0.0f, //顶点
    };
    
    //2.绘图索引
    GLuint indices[] =
    {
        0, 3, 2,
        0, 1, 3,
        0, 2, 4,
        0, 4, 1,
        2, 3, 4,
        1, 4, 3,
    };
    
    //顶点个数
    self.count = sizeof(indices) / sizeof(GLuint);
    
//    开辟顶点缓存区
    //将顶点数组放入数组缓冲区中 GL_ARRAY_BUFFER
    GLuint buffer;
    glGenBuffers(1, &buffer);
    glBindBuffer(GL_ARRAY_BUFFER, buffer);
    glBufferData(GL_ARRAY_BUFFER, sizeof(attrArr), attrArr, GL_STATIC_DRAW);
    
//    开辟索引缓存区
    //将索引数组存储到索引缓冲区 GL_ELEMENT_ARRAY_BUFFER
    GLuint index;
    glGenBuffers(1, &index);
    glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, index);
    glBufferData(GL_ELEMENT_ARRAY_BUFFER, sizeof(indices), indices, GL_STATIC_DRAW);
   
//    打开通道
    //使用顶点数据
    glEnableVertexAttribArray(GLKVertexAttribPosition);
    glVertexAttribPointer(GLKVertexAttribPosition, 3, GL_FLOAT, GL_FALSE, sizeof(GLfloat)*6, NULL);
    
    //使用颜色数据
    glEnableVertexAttribArray(GLKVertexAttribColor);
    glVertexAttribPointer(GLKVertexAttribColor, 3, GL_FLOAT, GL_FALSE, sizeof(GLfloat)*6, (GLfloat*)NULL+3);
    
    //着色器
    self.mEffect = [[GLKBaseEffect alloc] init];
    
    //投影视图
    CGSize size = self.view.bounds.size;
    float aspect = fabs(size.width / size.height);
    
    GLKMatrix4 projectionMatrix = GLKMatrix4MakePerspective(GLKMathDegreesToRadians(90), aspect, 0.1, 100);
    projectionMatrix = GLKMatrix4Scale(projectionMatrix, 1.0f, 1.0f, 1.0f);
    self.mEffect.transform.projectionMatrix = projectionMatrix;
    
    //模型视图
    GLKMatrix4 modelViewMatrix = GLKMatrix4Translate(GLKMatrix4Identity, 0, 0, -2.0);
    self.mEffect.transform.modelviewMatrix = modelViewMatrix;
    
//    定时器 GCD
    double seconds = 0.1;
    timer = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, dispatch_get_main_queue());
    dispatch_source_set_timer(timer, DISPATCH_TIME_NOW, seconds * NSEC_PER_SEC, 0.0);
    dispatch_source_set_event_handler(timer, ^{
        self.xDegree += 0.1 * self.bX;
        self.yDegree += 0.1 * self.bY;
        self.zDegree += 0.1 * self.bZ;
    });
    
    dispatch_resume(timer);
}

- (IBAction)xClick:(id)sender {
    _bX = !_bX;
}
- (IBAction)yClick:(id)sender {
    _bY = !_bY;
}
- (IBAction)zClick:(id)sender {
    _bZ = !_bZ;
}

//场景数据变化
- (void)update{
    
    GLKMatrix4 modelViewMatrix = GLKMatrix4Translate(GLKMatrix4Identity, 0, 0, -2.5);
    modelViewMatrix = GLKMatrix4RotateX(modelViewMatrix, self.xDegree);
    modelViewMatrix = GLKMatrix4RotateY(modelViewMatrix, self.yDegree);
    modelViewMatrix = GLKMatrix4RotateZ(modelViewMatrix, self.zDegree);
    
    self.mEffect.transform.modelviewMatrix = modelViewMatrix;
}

- (void)glkView:(GLKView *)view drawInRect:(CGRect)rect{
    glClearColor(0.3, 0.3, 0.3, 1.0);
    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
    
//    准备绘制
    [self.mEffect prepareToDraw];
    glDrawElements(GL_TRIANGLES, self.count, GL_UNSIGNED_INT, 0);
}

@end

File: 11_02_GLKit_三角形变换+纹理与颜色混合_Swift/GLKit_demo/CustomViewController.swift
//
//  CustomViewController.swift
//  GLKit_demo
//
//  Created by 陈嘉琳 on 2020/8/2.
//

import UIKit
import GLKit
import OpenGLES

class CustomViewController: GLKViewController {

    var mContext: EAGLContext!
    var mEffect: GLKBaseEffect!
    
    var count: Int = 0
    
    var xDegree: Float = 0
    var yDegree: Float = 0
    var zDegree: Float = 0
     
    var bX: Bool = false
    var bY: Bool = false
    var bZ: Bool = false
        
        var timer: DispatchSourceTimer!
//    var timer: Timer!

     override func viewDidLoad() {
         super.viewDidLoad()
         
         self.view.backgroundColor = UIColor.black
//        !!!!!必须设置代理，否则不执行
        self.delegate = self
         
//         新建图层
        setupContext()
        
//        渲染图层
        render()
     }
    
    override func glkView(_ view: GLKView, drawIn rect: CGRect) {
        glClearColor(0.3, 0.3, 0.3, 1.0)
        glClear(GLbitfield(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT))
        
        mEffect.prepareToDraw()
        glDrawElements(GLenum(GL_TRIANGLES), GLsizei(self.count), GLenum(GL_UNSIGNED_INT), UnsafePointer(bitPattern: 0))
    }
    
    @IBAction func xClick(_ sender: Any) {
        bX = !bX
    }
    
    @IBAction func yClick(_ sender: Any) {
        bY = !bY
        
    }
    
    @IBAction func zClick(_ sender: Any) {
        bZ = !bZ

    }
    
    @objc func reDegree(){
        self.xDegree += 0.1 * (self.bX ? 1 : 0)
        self.yDegree += 0.1 * (self.bY ? 1 : 0)
        self.zDegree += 0.1 * (self.bZ ? 1 : 0)
    }

}

 extension CustomViewController{
     
     fileprivate func setupContext(){
        
        self.mContext = EAGLContext(api: .openGLES2)
        
        let view: GLKView = self.view as! GLKView
        
        view.context = self.mContext
        view.drawableColorFormat = .RGBA8888
        view.drawableDepthFormat = .format24
        
        EAGLContext.setCurrent(self.mContext)
 
        glEnable(GLenum(GL_DEPTH_TEST))
     }
     
    
    fileprivate func render(){
        let attrArr: [GLfloat] = [
            -0.5, 0.5, 0.0,      1.0, 0.0, 1.0,         0.0, 1.0,//左上0
            0.5, 0.5, 0.0,       1.0, 0.0, 1.0,         1.0, 1.0,//右上1
            -0.5, -0.5, 0.0,     1.0, 1.0, 1.0,         0.0, 0.0,//左下2
            
            0.5, -0.5, 0.0,      1.0, 1.0, 1.0,         1.0, 0.0,//右下3
            0.0, 0.0, 1.0,       0.0, 1.0, 0.0,         0.5, 0.5//顶点4
        ]
        
        let indices: [GLuint] = [
                    0, 3, 2,
                   0, 1, 3,
                   0, 2, 4,
                   0, 4, 1,
                   2, 3, 4,
                   1, 4, 3,
        ]
        
        self.count = MemoryLayout<GLuint>.size*indices.count / MemoryLayout<GLuint>.size
        
//        开辟顶点缓存区
        var buffer: GLuint = GLuint()
        glGenBuffers(1, &buffer)
        glBindBuffer(GLenum(GL_ARRAY_BUFFER), buffer)
        glBufferData(GLenum(GL_ARRAY_BUFFER), MemoryLayout<GLfloat>.size*attrArr.count, attrArr, GLenum(GL_STATIC_DRAW))
        
//        开辟索引缓存区
        var index: GLuint = GLuint()
        glGenBuffers(1, &index)
        glBindBuffer(GLenum(GL_ELEMENT_ARRAY_BUFFER), index)
        glBufferData(GLenum(GL_ELEMENT_ARRAY_BUFFER), MemoryLayout<GLfloat>.size*indices.count, indices, GLenum(GL_STATIC_DRAW))
        
//        使用顶点
        glEnableVertexAttribArray(GLuint(GLKVertexAttrib.position.rawValue))
        glVertexAttribPointer(GLuint(GLKVertexAttrib.position.rawValue), 3, GLenum(GL_FLOAT), GLboolean(GL_FALSE), GLsizei(MemoryLayout<GLfloat>.size*8), UnsafeMutablePointer(bitPattern: 0))
        
//        使用索引
        glEnableVertexAttribArray(GLuint(GLKVertexAttrib.color.rawValue))
        glVertexAttribPointer(GLuint(GLKVertexAttrib.color.rawValue), 3, GLenum(GL_FLOAT), GLboolean(GL_FALSE), Int32(MemoryLayout<GLfloat>.size)*8, UnsafeMutablePointer(bitPattern: MemoryLayout<GLfloat>.size*3))
        
//        -------使用纹理
        glEnableVertexAttribArray(GLuint(GLKVertexAttrib.texCoord0.rawValue))
        glVertexAttribPointer(GLuint(GLKVertexAttrib.texCoord0.rawValue), 2, GLenum(GL_FLOAT), GLboolean(GL_FALSE), Int32(MemoryLayout<GLfloat>.size)*8, UnsafeMutablePointer(bitPattern: MemoryLayout<GLfloat>.size*6))
        
//        获取纹理路径
        let filePath = Bundle.main.path(forResource: "mouse", ofType: "jpg")
        guard let info: GLKTextureInfo = try? GLKTextureLoader .texture(withContentsOfFile: filePath!, options: [GLKTextureLoaderOriginBottomLeft: NSNumber(integerLiteral: 1)]) else {
            return
        }
        
//        着色器
        mEffect = GLKBaseEffect()
        mEffect.texture2d0.enabled = GLboolean(GL_TRUE)
        mEffect.texture2d0.name = info.name
        
//        投影矩阵
        let size = self.view.bounds.size
        let aspect = fabs(size.width / size.height)
        var projectionMatrix = GLKMatrix4MakePerspective(GLKMathDegreesToRadians(90), Float(aspect), 0.1, 100.0)
        projectionMatrix = GLKMatrix4Scale(projectionMatrix, 1.0, 1, 1)
        mEffect.transform.projectionMatrix = projectionMatrix
        
//        模型视图矩阵
        var modelViewMatrix = GLKMatrix4Translate(GLKMatrix4Identity, 0, 0, -2.0)
        mEffect.transform.modelviewMatrix = modelViewMatrix
        
//        设置定时器
//        https://www.jianshu.com/p/fc04be41c698
        let seconds: Double = 0.1
        timer = DispatchSource.makeTimerSource()
        //循环执行，马上开始，间隔为1s,误差允许10微秒
        timer?.schedule(deadline: DispatchTime.now(), repeating: seconds, leeway: .milliseconds(10))
        timer.setEventHandler {
            self.xDegree += 0.1 * (self.bX ? 1 : 0)
            self.yDegree += 0.1 * (self.bY ? 1 : 0)
            self.zDegree += 0.1 * (self.bZ ? 1 : 0)
        }
        timer.resume()

    }

 }


extension CustomViewController: GLKViewControllerDelegate{
    func glkViewControllerUpdate(_ controller: GLKViewController) {
        var modelViewMatrix = GLKMatrix4Translate(GLKMatrix4Identity, 0, 0, -2.5)
        modelViewMatrix = GLKMatrix4RotateX(modelViewMatrix, xDegree)
        modelViewMatrix = GLKMatrix4RotateY(modelViewMatrix, yDegree)
        modelViewMatrix = GLKMatrix4RotateZ(modelViewMatrix, zDegree)

        mEffect.transform.modelviewMatrix = modelViewMatrix
    }
    
}
